File: src/daemon.ts
#!/usr/bin/env node

/**
 * clangd-daemon - Background server for clangd-query
 *
 * This daemon maintains a persistent clangd instance to provide fast code intelligence
 * queries without repeated indexing overhead. Features:
 * - Single daemon per project root via lock files
 * - Unix domain socket communication with JSON-RPC 2.0 protocol
 * - Automatic idle timeout after 30 minutes of inactivity
 * - Graceful shutdown with proper resource cleanup
 * - Concurrent client handling
 *
 * The daemon is automatically started by clangd-query when needed and runs in the
 * background until explicitly stopped or idle timeout expires.
 */

import * as net from "node:net";
import * as fs from "node:fs";
import * as path from "node:path";
import * as os from "node:os";
import { ClangdClient } from "./clangd-client.js";
import { Logger } from "./logger.js";
import {
  generateSocketPath,
  generateLockFilePath,
  getLogFilePath,
  readLockFile,
  writeLockFile,
  cleanupStaleLockFile,
  calculateBuildTimestamp,
  type LockFileData,
} from "./socket-utils.js";
import * as commands from "./commands/index.js";
import { FileWatcher, FileEvent } from "./file-watcher.js";

// Types for JSON-RPC 2.0
interface JsonRpcRequest {
  jsonrpc: "2.0";
  id: string | number;
  method: string;
  params?: any;
}

interface JsonRpcResponse {
  jsonrpc: "2.0";
  id: string | number;
  result?: any;
  error?: {
    code: number;
    message: string;
    data?: any;
  };
  logs?: string[]; // Request-specific debug logs
}

interface DaemonStatus {
  uptime: number;
  projectRoot: string;
  memory: {
    heapUsed: string;
    heapTotal: string;
    rss: string;
  };
  requestCount: number;
  lastRequestTime: number;
  indexingComplete: boolean;
}

enum LogLevel {
  ERROR = 0,
  INFO = 1,
  DEBUG = 2,
}

/**
 * Logger implementation that captures logs to a buffer during request processing.
 * This allows the daemon to capture logs from ClangdClient and include them in responses.
 */
class RequestLogger implements Logger {
  constructor(private buffer: string[]) {}

  error(message: string, ...args: any[]): void {
    this.captureLog(LogLevel.ERROR, message, args);
  }

  info(message: string, ...args: any[]): void {
    this.captureLog(LogLevel.INFO, message, args);
  }

  debug(message: string, ...args: any[]): void {
    this.captureLog(LogLevel.DEBUG, message, args);
  }

  private captureLog(level: LogLevel, message: string, args: any[]): void {
    const levelName = LogLevel[level];
    let logMessage = `[${levelName}] ${message}`;

    // Add arguments if any
    if (args.length > 0) {
      // Format each argument nicely
      for (const arg of args) {
        if (typeof arg === 'string') {
          // Try to parse as JSON for better formatting
          try {
            const parsed = JSON.parse(arg);
            logMessage += "\n" + JSON.stringify(parsed, null, 2);
          } catch {
            // Not JSON, just append as-is
            logMessage += "\n" + arg;
          }
        } else {
          logMessage += "\n" + JSON.stringify(arg, null, 2);
        }
      }
    }

    // Always capture to buffer
    this.buffer.push(logMessage);
  }
}

interface LogEntry {
  level: LogLevel;
  timestamp: string;
  message: string;
}

/**
 * Logger implementation for the daemon's internal operations.
 * Writes to the daemon's log file and maintains recent logs in memory.
 */
class DaemonLogger implements Logger {
  private logStream: fs.WriteStream;
  private recentLogs: LogEntry[] = [];
  private readonly maxRecentLogs = 1000;

  constructor(
    private logFilePath: string,
    private level: LogLevel
  ) {
    // Create log directory if needed
    const logDir = path.dirname(this.logFilePath);
    fs.mkdirSync(logDir, { recursive: true });

    // Delete log file if it's too large (> 1MB)
    const maxLogSize = 1024 * 1024; // 1MB
    try {
      const stats = fs.statSync(this.logFilePath);
      if (stats.size > maxLogSize) {
        fs.unlinkSync(this.logFilePath);
      }
    } catch {
      // File doesn't exist yet, that's fine
    }

    // Create log stream with append mode
    this.logStream = fs.createWriteStream(this.logFilePath, {
      flags: "a",
      encoding: "utf8",
    });
  }

  /**
   * Get filtered logs based on log level
   * @param requestedLevel The minimum log level to include ("error", "info", or "debug")
   * @param lines Maximum number of lines to return
   */
  getFilteredLogs(requestedLevel: string, lines: number): { logs: string[], totalCount: number } {
    // Filter logs based on requested level
    let filteredLogs = this.recentLogs;

    if (requestedLevel !== "debug") {
      const minLevel = requestedLevel === "error" ? LogLevel.ERROR : LogLevel.INFO;
      filteredLogs = this.recentLogs.filter(entry => entry.level <= minLevel);
    }

    // Get the last N lines from filtered logs
    const startIndex = Math.max(0, filteredLogs.length - lines);
    const selectedLogs = filteredLogs.slice(startIndex);

    // Format logs as strings for output
    const formattedLogs = selectedLogs.map(entry =>
      `[${entry.timestamp}] [${LogLevel[entry.level]}] ${entry.message}`
    );

    return {
      logs: formattedLogs,
      totalCount: filteredLogs.length
    };
  }

  /**
   * Close the log stream
   */
  close(): void {
    this.logStream.end();
  }

  error(message: string, ...args: any[]): void {
    this.log(LogLevel.ERROR, message, ...args);
  }

  info(message: string, ...args: any[]): void {
    this.log(LogLevel.INFO, message, ...args);
  }

  debug(message: string, ...args: any[]): void {
    this.log(LogLevel.DEBUG, message, ...args);
  }

  private log(level: LogLevel, message: string, ...args: any[]): void {
    const now = new Date();
    const timestamp = now.toTimeString().split(' ')[0]; // HH:MM:SS format
    const levelName = LogLevel[level];

    // Format the full message
    let fullMessage = message;
    if (args.length > 0) {
      const formattedArgs = args.map(arg =>
        typeof arg === 'object' ? JSON.stringify(arg) : String(arg)
      ).join(' ');
      fullMessage += ' ' + formattedArgs;
    }

    // Add to recent logs
    this.addToRecentLogs({
      level: level,
      timestamp: timestamp,
      message: fullMessage
    });

    // Only write to file if the message level is at or below the configured level
    if (level <= this.level) {
      // Format as "Aug 14 15:13:12.080" using locale string
      const dateStr = now.toLocaleDateString('en-US', { month: 'short', day: '2-digit' });
      const timeStr = now.toLocaleTimeString('en-US', { hour12: false, hour: '2-digit', minute: '2-digit', second: '2-digit' });
      const ms = now.getMilliseconds().toString().padStart(3, '0');
      const fullTimestamp = `${dateStr} ${timeStr}.${ms}`;
      let fileLogMessage = `[${fullTimestamp}] [${levelName}] ${message}`;
      if (args.length > 0) {
        const formattedArgs = args.map(arg =>
          typeof arg === 'object' ? JSON.stringify(arg) : String(arg)
        ).join(' ');
        fileLogMessage += ' ' + formattedArgs;
      }

      this.logStream.write(fileLogMessage + "\n");
      // Force flush to ensure logs are immediately visible on disk
      this.logStream.cork();
      this.logStream.uncork();
    }
  }

  private addToRecentLogs(entry: LogEntry): void {
    this.recentLogs.push(entry);

    // Maintain circular buffer size
    if (this.recentLogs.length > this.maxRecentLogs) {
      this.recentLogs.shift();
    }
  }
}

class ClangdDaemon {
  private projectRoot: string;
  private socketPath: string;
  private lockFilePath: string;
  private logFilePath: string;
  private server: net.Server | null = null;
  private clangdClient: ClangdClient | null = null;
  private startTime: number = Date.now();
  private requestCount: number = 0;
  private lastRequestTime: number = Date.now();
  private idleTimer: NodeJS.Timeout | null = null;
  private idleTimeoutMs: number;
  private connections: Set<net.Socket> = new Set();
  private logLevel: LogLevel;
  private requestLogBuffer: string[] | null = null; // Buffer for request-specific logs
  private fileWatcher: FileWatcher | null = null;
  private logger!: DaemonLogger; // Daemon-wide logger instance (initialized in constructor)

  constructor(projectRoot: string) {
    this.projectRoot = path.resolve(projectRoot);
    this.socketPath = generateSocketPath(this.projectRoot);
    this.lockFilePath = generateLockFilePath(this.projectRoot);
    this.logFilePath = getLogFilePath(this.projectRoot);

    // Get idle timeout from environment or use default (30 minutes)
    const timeoutSeconds = parseInt(process.env.CLANGD_DAEMON_TIMEOUT || "1800", 10);
    this.idleTimeoutMs = timeoutSeconds * 1000;

    // Set default log level to INFO
    // This captures important events without too much noise
    this.logLevel = LogLevel.INFO;

    // Create the daemon-wide logger (non-null from this point on)
    this.logger = new DaemonLogger(this.logFilePath, this.logLevel);
  }




  private initializeLogging(): void {
    this.logger.info(`Daemon starting for project: ${this.projectRoot}`);
    this.logger.info(`PID: ${process.pid}`);
    this.logger.info(`Socket path: ${this.socketPath}`);
    this.logger.info(`Idle timeout: ${this.idleTimeoutMs / 1000} seconds`);
  }

  private resetIdleTimer(): void {
    if (this.idleTimer) {
      clearTimeout(this.idleTimer);
    }

    this.idleTimer = setTimeout(() => {
      this.logger.info("Idle timeout reached, shutting down");
      this.shutdown().catch((error) => {
        this.logger.error("Error during idle shutdown", error);
        process.exit(1);
      });
    }, this.idleTimeoutMs);
  }

  private async checkExistingDaemon(): Promise<boolean> {
    // Clean up stale lock files first
    const wasStale = cleanupStaleLockFile(this.lockFilePath);
    if (wasStale) {
      this.logger.info("Cleaned up stale lock file");
    }

    // Check if lock file still exists
    const lockData = readLockFile(this.lockFilePath);
    if (lockData) {
      this.logger.error(`Another daemon is already running (PID: ${lockData.pid})`);
      return true;
    }

    return false;
  }

  private createLockFile(): void {
    const lockData: LockFileData = {
      pid: process.pid,
      socketPath: this.socketPath,
      startTime: this.startTime,
      projectRoot: this.projectRoot,
      buildTimestamp: calculateBuildTimestamp(import.meta.url),
    };

    writeLockFile(this.lockFilePath, lockData);
    this.logger.info("Lock file created");
  }

  private removeLockFile(): void {
    try {
      fs.unlinkSync(this.lockFilePath);
      this.logger.info("Lock file removed");
    } catch (error) {
      this.logger.error("Failed to remove lock file", error);
    }
  }

  private removeSocketFile(): void {
    try {
      if (fs.existsSync(this.socketPath)) {
        fs.unlinkSync(this.socketPath);
        this.logger.info("Socket file removed");
      }
    } catch (error) {
      this.logger.error("Failed to remove socket file", error);
    }
  }

  private async initializeClangd(): Promise<void> {
    this.logger.info("Initializing clangd");

    this.clangdClient = new ClangdClient(this.projectRoot, {
      clangdPath: process.env.CLANGD_PATH,
      logger: this.logger!,
    });

    await this.clangdClient.start();
    this.logger.info("Clangd initialized successfully");
  }

  /**
   * Handle file change events from the file watcher
   */
  private async handleFileChanges(changes: FileEvent[]): Promise<void> {
    if (!this.clangdClient) {
      this.logger.error("Cannot handle file changes: clangd client not initialized");
      return;
    }

    try {
      // Convert our FileEvent type to LSP FileEvent type
      const lspFileEvents = changes.map(change => ({
        uri: change.uri,
        type: change.type,
      }));

      await this.clangdClient.sendFileChangeNotification(lspFileEvents);

      this.logger.info(`Notified clangd about ${changes.length} file changes`);
      this.logger.debug("File change details:", lspFileEvents);

      // Check if compile_commands.json changed
      const hasCompileCommandsChanged = changes.some(
        change => change.uri.endsWith("/compile_commands.json")
      );

      if (hasCompileCommandsChanged) {
        this.logger.info("compile_commands.json changed - clangd should reindex the project");
        // clangd should handle this automatically when notified
      }
    } catch (error) {
      this.logger.error("Failed to notify clangd about file changes", error);
    }
  }

  /**
   * Initialize the file watcher
   */
  private async initializeFileWatcher(): Promise<void> {
    this.logger.info("Initializing file watcher");

    this.fileWatcher = new FileWatcher({
      projectRoot: this.projectRoot,
      onFileChanges: (changes) => this.handleFileChanges(changes),
      logger: {
        error: (msg, ...args) => this.logger.error(msg, ...args),
        info: (msg, ...args) => this.logger.info(msg, ...args),
        debug: (msg, ...args) => this.logger.debug(msg, ...args),
      },
      debounceMs: 500,
    });

    await this.fileWatcher.start();
    this.logger.info("File watcher initialized successfully");
  }

  private async handleRequest(request: JsonRpcRequest): Promise<JsonRpcResponse> {
    this.requestCount++;
    this.lastRequestTime = Date.now();
    this.resetIdleTimer();

    // Initialize request log buffer to capture all logs during this request
    this.requestLogBuffer = [];
    const requestLogger = new RequestLogger(this.requestLogBuffer);

    const response: JsonRpcResponse = {
      jsonrpc: "2.0",
      id: request.id,
    };

    try {
      switch (request.method) {
        case "searchSymbols": {
          const { query, limit } = request.params || {};
          if (!query) {
            throw new Error("Missing required parameter: query");
          }
          const result = await commands.searchSymbolsAsText(this.clangdClient!, query, limit, requestLogger);
          response.result = { text: result };
          break;
        }

        case "viewSourceCode": {
          const { query } = request.params || {};
          if (!query) {
            throw new Error("Missing required parameter: query");
          }
          const result = await commands.viewSourceCodeAsText(this.clangdClient!, query, requestLogger);
          response.result = { text: result };
          break;
        }

        case "findReferences": {
          const { location } = request.params || {};
          if (!location) {
            throw new Error("Missing required parameter: location");
          }
          const result = await commands.findReferencesAsText(this.clangdClient!, location, requestLogger);
          response.result = { text: result };
          break;
        }

        case "findReferencesToSymbol": {
          const { symbolName } = request.params || {};
          if (!symbolName) {
            throw new Error("Missing required parameter: symbolName");
          }
          const result = await commands.findReferencesToSymbolAsText(this.clangdClient!, symbolName, requestLogger);
          response.result = { text: result };
          break;
        }

        case "getTypeHierarchy": {
          const { className } = request.params || {};
          if (!className) {
            throw new Error("Missing required parameter: className");
          }
          const result = await commands.getTypeHierarchyAsText(this.clangdClient!, className, requestLogger);
          response.result = { text: result };
          break;
        }

        case "getSignature": {
          const { functionName } = request.params || {};
          if (!functionName) {
            throw new Error("Missing required parameter: functionName");
          }
          const result = await commands.getSignatureAsText(this.clangdClient!, functionName, requestLogger);
          response.result = { text: result };
          break;
        }

        case "getInterface": {
          const { className } = request.params || {};
          if (!className) {
            throw new Error("Missing required parameter: className");
          }
          const result = await commands.getInterfaceAsText(this.clangdClient!, className, requestLogger);
          response.result = { text: result };
          break;
        }

        case "getShow": {
          const { symbol } = request.params || {};
          if (!symbol) {
            throw new Error("Missing required parameter: symbol");
          }
          const result = await commands.getShowAsText(this.clangdClient!, symbol, requestLogger);
          response.result = { text: result };
          break;
        }

        case "getContext": {
          // Keep for backward compatibility but redirect to show
          const { symbol } = request.params || {};
          if (!symbol) {
            throw new Error("Missing required parameter: symbol");
          }
          const result = await commands.getShowAsText(this.clangdClient!, symbol, requestLogger);
          response.result = { text: result };
          break;
        }

        case "ping": {
          response.result = { status: "ok", timestamp: Date.now() };
          break;
        }

        case "getStatus": {
          const memUsage = process.memoryUsage();
          const status: DaemonStatus = {
            uptime: Date.now() - this.startTime,
            projectRoot: this.projectRoot,
            memory: {
              heapUsed: `${Math.round(memUsage.heapUsed / 1024 / 1024)}MB`,
              heapTotal: `${Math.round(memUsage.heapTotal / 1024 / 1024)}MB`,
              rss: `${Math.round(memUsage.rss / 1024 / 1024)}MB`,
            },
            requestCount: this.requestCount,
            lastRequestTime: this.lastRequestTime,
            indexingComplete: true, // We wait for indexing during startup
          };
          response.result = status;
          break;
        }

        case "getRecentLogs": {
          const lines = request.params?.lines || 100;
          const requestedLevel = request.params?.logLevel || "debug";

          // Get filtered logs from logger
          const result = this.logger.getFilteredLogs(requestedLevel, lines);

          response.result = {
            logs: result.logs,
            totalCount: result.totalCount,
            returnedCount: result.logs.length
          };
          break;
        }

        case "shutdown": {
          response.result = { status: "shutting down" };
          // Schedule shutdown after sending response
          setTimeout(() => {
            this.shutdown().catch((error) => {
              this.logger.error("Error during shutdown", error);
              process.exit(1);
            });
          }, 100);
          break;
        }

        default:
          response.error = {
            code: -32601,
            message: `Method not found: ${request.method}`,
          };
      }
    } catch (error) {
      this.logger.error(`Error handling request ${request.method}`, error);
      response.error = {
        code: -32000,
        message: error instanceof Error ? error.message : "Unknown error",
        data: error instanceof Error ? { stack: error.stack } : undefined,
      };
    }

    // Include captured logs in the response
    if (this.requestLogBuffer !== null && this.requestLogBuffer.length > 0) {
      response.logs = this.requestLogBuffer;
    }

    // Clear request log buffer
    this.requestLogBuffer = null;

    return response;
  }

  private handleConnection(socket: net.Socket): void {
    this.logger.debug("New client connected");
    this.connections.add(socket);

    let buffer = "";

    socket.on("data", async (data) => {
      buffer += data.toString();

      // Process complete JSON messages (newline-delimited)
      let lines = buffer.split("\n");
      buffer = lines.pop() || ""; // Keep incomplete line in buffer

      for (const line of lines) {
        if (!line.trim()) continue;

        try {
          const request = JSON.parse(line) as JsonRpcRequest;
          this.logger.debug(`Received request: ${request.method}`, request.params);

          const response = await this.handleRequest(request);
          socket.write(JSON.stringify(response) + "\n");

          this.logger.debug(`Sent response for request ${request.id}`);
        } catch (error) {
          this.logger.error("Failed to parse request", error);
          const errorResponse: JsonRpcResponse = {
            jsonrpc: "2.0",
            id: 0,
            error: {
              code: -32700,
              message: "Parse error",
            },
          };
          socket.write(JSON.stringify(errorResponse) + "\n");
        }
      }
    });

    socket.on("error", (error) => {
      this.logger.error("Socket error", error);
    });

    socket.on("close", () => {
      this.logger.debug("Client disconnected");
      this.connections.delete(socket);
    });
  }

  async start(): Promise<void> {
    // Initialize logging first
    this.initializeLogging();

    // Check for existing daemon
    if (await this.checkExistingDaemon()) {
      throw new Error("Another daemon is already running");
    }

    // Create lock file
    this.createLockFile();

    // Clean up socket file if it exists
    this.removeSocketFile();

    // Initialize clangd
    await this.initializeClangd();

    // Initialize file watcher after clangd is ready
    await this.initializeFileWatcher();

    // Create socket server
    this.server = net.createServer((socket) => this.handleConnection(socket));

    // Set socket permissions (owner read/write only)
    this.server.listen(this.socketPath, () => {
      fs.chmodSync(this.socketPath, 0o600);
      this.logger.info(`Daemon listening on ${this.socketPath}`);
    });

    // Start idle timer
    this.resetIdleTimer();

    // Handle graceful shutdown
    process.on("SIGTERM", () => {
      this.logger.info("Received SIGTERM, shutting down");
      this.shutdown().catch((error) => {
        this.logger.error("Error during shutdown", error);
        process.exit(1);
      });
    });

    process.on("SIGINT", () => {
      this.logger.info("Received SIGINT, shutting down");
      this.shutdown().catch((error) => {
        this.logger.error("Error during shutdown", error);
        process.exit(1);
      });
    });

    // Handle uncaught errors
    process.on("uncaughtException", (error) => {
      this.logger.error("Uncaught exception", error);
      this.shutdown().then(() => {
        process.exit(1);
      });
    });

    process.on("unhandledRejection", (reason, promise) => {
      this.logger.error("Unhandled rejection", reason);
      this.shutdown().then(() => {
        process.exit(1);
      });
    });
  }

  async shutdown(): Promise<void> {
    this.logger.info("Starting shutdown sequence");

    // Clear idle timer
    if (this.idleTimer) {
      clearTimeout(this.idleTimer);
      this.idleTimer = null;
    }

    // Close all connections
    for (const socket of this.connections) {
      socket.destroy();
    }
    this.connections.clear();

    // Close server
    if (this.server) {
      await new Promise<void>((resolve) => {
        this.server!.close(() => {
          this.logger.info("Server closed");
          resolve();
        });
      });
    }

    // Stop file watcher
    if (this.fileWatcher) {
      await this.fileWatcher.stop();
      this.logger.info("File watcher stopped");
    }

    // Stop clangd
    if (this.clangdClient) {
      await this.clangdClient.stop();
      this.logger.info("Clangd stopped");
    }

    // Remove lock file and socket
    this.removeLockFile();
    this.removeSocketFile();

    // Log final message before closing logger
    this.logger.info("Shutdown complete");

    // Close logger
    this.logger.close();

    process.exit(0);
  }
}

// Main entry point
async function main() {
  const projectRoot = process.argv[2];

  if (!projectRoot) {
    console.error("Usage: clangd-daemon <project-root>");
    process.exit(1);
  }

  const daemon = new ClangdDaemon(projectRoot);

  try {
    await daemon.start();
  } catch (error) {
    console.error("Failed to start daemon:", error);
    process.exit(1);
  }
}

// Run if executed directly
if (import.meta.url === `file://${process.argv[1]}`) {
  main().catch((error) => {
    console.error("Fatal error:", error);
    process.exit(1);
  });
}

File: go/internal/daemon/daemon.go
package daemon

import (
	"encoding/json"
	"fmt"
	"net"
	"os"
	"os/signal"
	"sync"
	"syscall"
	"time"

	"clangd-query/internal/clangd"
	"clangd-query/internal/commands"
	"clangd-query/internal/logger"
)

// Config contains daemon configuration
type Config struct {
	ProjectRoot string
	Verbose     bool
}

// Daemon manages the clangd process and handles client connections
type Daemon struct {
	projectRoot   string
	socketPath    string
	logger        logger.Logger
	clangdClient  *clangd.ClangdClient
	fileWatcher   *FileWatcher
	listener      net.Listener
	idleTimer     *time.Timer
	idleTimeout   time.Duration
	mu            sync.Mutex
	shutdown      chan struct{}
	connections   int
	totalRequests int
	startTime     time.Time
}

// Request represents a client request
type Request struct {
	ID     interface{}            `json:"id"`
	Method string                 `json:"method"`
	Params map[string]interface{} `json:"params,omitempty"`
}

// Response represents a daemon response
type Response struct {
	ID     interface{}     `json:"id"`
	Result json.RawMessage `json:"result,omitempty"`
	Error  *ErrorResponse  `json:"error,omitempty"`
}

// ErrorResponse represents an error response
type ErrorResponse struct {
	Code    int    `json:"code"`
	Message string `json:"message"`
}

// Run starts the daemon
func Run(config *Config) {
	daemon := &Daemon{
		projectRoot: config.ProjectRoot,
		socketPath:  GetSocketPath(config.ProjectRoot),
		shutdown:    make(chan struct{}),
		startTime:   time.Now(),
	}

	// Setup logging with config
	if err := daemon.setupLogging(config.Verbose); err != nil {
		fmt.Fprintf(os.Stderr, "Failed to setup logging: %v\n", err)
		os.Exit(1)
	}
	defer func() {
		if fileLogger, ok := daemon.logger.(*logger.FileLogger); ok {
			fileLogger.Close()
		}
	}()

	daemon.logger.Info("Starting daemon for project: %s", config.ProjectRoot)

	// Check for existing daemon
	if err := daemon.checkExistingDaemon(); err != nil {
		daemon.logger.Error("Error checking existing daemon: %v", err)
		os.Exit(1)
	}

	// Write lock file
	if err := WriteLockFile(config.ProjectRoot, os.Getpid(), daemon.socketPath); err != nil {
		daemon.logger.Error("Failed to write lock file: %v", err)
		os.Exit(1)
	}
	defer RemoveLockFile(config.ProjectRoot)

	// Ensure compilation database exists
	buildDir, err := EnsureCompilationDatabase(config.ProjectRoot, daemon.logger)
	if err != nil {
		daemon.logger.Error("Failed to find compilation database: %v", err)
		os.Exit(1)
	}

	// Start clangd
	daemon.logger.Info("Starting clangd with build directory: %s", buildDir)
	daemon.clangdClient, err = clangd.NewClangdClient(config.ProjectRoot, buildDir, daemon.logger)
	if err != nil {
		daemon.logger.Error("Failed to start clangd: %v", err)
		os.Exit(1)
	}
	defer daemon.clangdClient.Stop()

	// Setup file watcher
	daemon.fileWatcher, err = NewFileWatcher(config.ProjectRoot, daemon.onFilesChanged, daemon.logger)
	if err != nil {
		daemon.logger.Error("Failed to setup file watcher: %v", err)
		// Continue without file watching
	} else {
		defer daemon.fileWatcher.Stop()
	}

	// Setup idle timeout
	daemon.setupIdleTimeout()

	// Setup signal handlers
	daemon.setupSignalHandlers()

	// Start socket server
	if err := daemon.startSocketServer(); err != nil {
		daemon.logger.Error("Failed to start socket server: %v", err)
		os.Exit(1)
	}

	daemon.logger.Info("Daemon started successfully")

	// Wait for shutdown
	<-daemon.shutdown

	daemon.logger.Info("Daemon shutting down")
}

func (d *Daemon) setupLogging(verbose bool) error {
	logPath := GetLogPath(d.projectRoot)

	// Determine file log level based on verbose flag
	fileLogLevel := logger.LevelInfo
	if verbose {
		fileLogLevel = logger.LevelDebug
	}

	// Create file logger
	fileLogger, err := logger.NewFileLogger(logPath, fileLogLevel)
	if err != nil {
		return err
	}

	d.logger = fileLogger
	return nil
}

func (d *Daemon) checkExistingDaemon() error {
	lockInfo, err := ReadLockFile(d.projectRoot)
	if err != nil {
		return err
	}

	if lockInfo != nil {
		if IsProcessAlive(lockInfo.PID) {
			if IsDaemonStale(lockInfo) {
				d.logger.Info("Existing daemon is stale, attempting to stop it")
				// Try to gracefully stop the old daemon
				syscall.Kill(lockInfo.PID, syscall.SIGTERM)
				time.Sleep(100 * time.Millisecond)
			} else {
				return fmt.Errorf("daemon already running with PID %d", lockInfo.PID)
			}
		} else {
			d.logger.Debug("Found stale lock file, cleaning up")
		}

		// Clean up old socket
		CleanupSocket(lockInfo.SocketPath)
		RemoveLockFile(d.projectRoot)
	}

	return nil
}

func (d *Daemon) setupIdleTimeout() {
	// Get timeout from environment or use default (30 minutes)
	timeoutStr := os.Getenv("CLANGD_DAEMON_TIMEOUT")
	if timeoutStr == "" {
		d.idleTimeout = 30 * time.Minute
	} else {
		if timeout, err := time.ParseDuration(timeoutStr); err == nil {
			d.idleTimeout = timeout
		} else {
			d.idleTimeout = 30 * time.Minute
		}
	}

	d.resetIdleTimer()
}

func (d *Daemon) resetIdleTimer() {
	d.mu.Lock()
	defer d.mu.Unlock()

	if d.idleTimer != nil {
		d.idleTimer.Stop()
	}

	d.idleTimer = time.AfterFunc(d.idleTimeout, func() {
		d.logger.Info("Idle timeout reached, shutting down")
		close(d.shutdown)
	})
}

func (d *Daemon) setupSignalHandlers() {
	sigChan := make(chan os.Signal, 1)
	signal.Notify(sigChan, syscall.SIGTERM, syscall.SIGINT)

	go func() {
		sig := <-sigChan
		d.logger.Info("Received signal: %v", sig)
		close(d.shutdown)
	}()
}

func (d *Daemon) startSocketServer() error {
	// Remove old socket if it exists
	CleanupSocket(d.socketPath)

	// Create socket listener
	listener, err := net.Listen("unix", d.socketPath)
	if err != nil {
		return err
	}
	d.listener = listener

	// Start accepting connections
	go d.acceptConnections()

	return nil
}

func (d *Daemon) acceptConnections() {
	defer d.listener.Close()
	defer CleanupSocket(d.socketPath)

	for {
		conn, err := d.listener.Accept()
		if err != nil {
			select {
			case <-d.shutdown:
				return
			default:
				d.logger.Error("Error accepting connection: %v", err)
				continue
			}
		}

		d.resetIdleTimer()
		go d.handleConnection(conn)
	}
}

func (d *Daemon) handleConnection(conn net.Conn) {
	defer conn.Close()

	d.mu.Lock()
	d.connections++
	clientID := d.connections
	d.mu.Unlock()

	d.logger.Info("Client %d connected", clientID)

	decoder := json.NewDecoder(conn)
	encoder := json.NewEncoder(conn)

	for {
		var req Request
		if err := decoder.Decode(&req); err != nil {
			if err.Error() != "EOF" {
				d.logger.Error("Client %d: Error decoding request: %v", clientID, err)
			}
			break
		}

		d.mu.Lock()
		d.totalRequests++
		d.mu.Unlock()

		d.logger.Info("Client %d: Request %s", clientID, req.Method)

		// Handle the request
		result, err := d.handleRequest(req)

		// Send response
		resp := Response{
			ID: req.ID,
		}

		if err != nil {
			resp.Error = &ErrorResponse{
				Code:    -1,
				Message: err.Error(),
			}
		} else {
			resp.Result = result
		}

		if err := encoder.Encode(resp); err != nil {
			d.logger.Error("Client %d: Error encoding response: %v", clientID, err)
			break
		}
	}

	d.logger.Info("Client %d disconnected", clientID)
}

func (d *Daemon) handleRequest(req Request) (json.RawMessage, error) {
	switch req.Method {
	case "status":
		return d.handleStatus()
	case "logs":
		return d.handleLogs(req)
	case "shutdown":
		go func() {
			time.Sleep(100 * time.Millisecond)
			close(d.shutdown)
		}()
		return json.Marshal(map[string]string{"status": "shutting down"})
	}

	// All other commands go to clangd
	input, _ := req.Params["symbol"].(string)

	limit := -1
	if l, ok := req.Params["limit"].(float64); ok {
		limit = int(l)
	}

	var output string
	var err error

	switch req.Method {
	case "search":
		output, err = commands.Search(d.clangdClient, input, limit, d.logger)
	case "show":
		output, err = commands.Show(d.clangdClient, input, d.logger)
	case "view":
		output, err = commands.View(d.clangdClient, input, d.logger)
	case "usages":
		output, err = commands.Usages(d.clangdClient, input, limit, d.logger)
	case "hierarchy":
		output, err = commands.Hierarchy(d.clangdClient, input, limit, d.logger)
	case "signature":
		output, err = commands.Signature(d.clangdClient, input, d.logger)
	case "interface":
		output, err = commands.Interface(d.clangdClient, input, d.logger)
	default:
		return nil, fmt.Errorf("unknown method: %s", req.Method)
	}

	if err != nil {
		d.logger.Error("%s failed: %v", req.Method, err)
		return nil, err
	}

	return json.Marshal(map[string]string{"output": output})
}

func (d *Daemon) handleStatus() (json.RawMessage, error) {
	d.mu.Lock()
	defer d.mu.Unlock()

	status := map[string]interface{}{
		"pid":           os.Getpid(),
		"projectRoot":   d.projectRoot,
		"uptime":        time.Since(d.startTime).String(),
		"totalRequests": d.totalRequests,
		"connections":   d.connections,
		"idleTimeout":   d.idleTimeout.String(),
	}

	return json.Marshal(status)
}

func (d *Daemon) handleLogs(req Request) (json.RawMessage, error) {
	// Get log level from params (default to INFO and above)
	minLevel := logger.LevelInfo
	if level, ok := req.Params["level"].(string); ok {
		switch level {
		case "error":
			minLevel = logger.LevelError
		case "verbose", "debug":
			minLevel = logger.LevelDebug
		default:
			minLevel = logger.LevelInfo
		}
	}

	// Get filtered logs from memory
	logs := d.logger.GetLogs(minLevel)
	return json.Marshal(map[string]string{"logs": logs})
}

func (d *Daemon) onFilesChanged(files []string) {
	d.logger.Debug("Files changed: %v", files)

	if d.clangdClient != nil {
		// Notify clangd about file changes
		d.clangdClient.OnFilesChanged(files)
	}
}
